_posts/2021-01-11-zappa-serverless-docker.md

```diff
@@ -0,0 +1,224 @@
+```
+
+We'll also have a batch process we run at regular intervals:
+
+```python
+# zdf/process.py
+from datetime import datetime
+
+def run_process():
+    now_str = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
+    print(f"The current time is {now_str}")
+```
+
+## zappa_settings.json
+
+Our `zappa_settings.json` is pretty minimal, and doesn't require any changes from your typical one. The primary difference is you don't need to specify a Python runtime since that will be set in the Docker image.
+
+With this settings configuration, we get the following:
+
+* Flask app in `zdf/app.py` will be served
+* `process.run_process` function will run every 2 hours
+* The environment variable `EXAMPLE_ENV_VAR` will be available in the Docker image
+
+```json
+{
+    "lambda_docker_flask": {
+        "app_function": "zdf.app.app",
+        "project_name": "test",
+        "s3_bucket": "lambda-docker-flask",
+        "environment_variables": {
+            "EXAMPLE_ENV_VAR": "prod"
+        },
+        "events": [
+            {
+               "function": "zdf.process.run_process",
+               "expression": "cron(0 */2 * * ? *)"
+            }
+        ],
+        "lambda_description": "Zappa + Docker + Flask"
+    }
+}
+```
+
+## Building the Docker Image
+
+### Dockerfile
+
+As mentioned above, you must provide a prebuilt Docker image to Zappa. This Docker image must be built according to the [standards outlined by AWS](https://docs.aws.amazon.com/lambda/latest/dg/images-create.html). Here's the Dockerfile I used for the demo above:
+
+
+```bash
+FROM amazon/aws-lambda-python:3.8
+
+ARG FUNCTION_DIR="/var/task/"
+
+COPY ./ ${FUNCTION_DIR}
+
+# Setup Python environment
+RUN pip install poetry
+RUN POETRY_VIRTUALENVS_CREATE=false poetry install --no-root
+
+# Grab the zappa handler.py and put it in the working directory
+RUN ZAPPA_HANDLER_PATH=$( \
+    python -c "from zappa import handler; print (handler.__file__)" \
+    ) \
+    && echo $ZAPPA_HANDLER_PATH \
+    && cp $ZAPPA_HANDLER_PATH ${FUNCTION_DIR}
+
+
+CMD [ "handler.lambda_handler" ]
+```
+
+The Dockerfile is pretty straightforward. I'm building off a base image provided by AWS, but you can implement yours using a [different base image](https://docs.aws.amazon.com/lambda/latest/dg/images-create.html#images-create-2) if you'd like. I copy my application code into the image and setup my Python environment with [Poetry](https://python-poetry.org/). 
+
+The **only Zappa specific steps** are the last two. The Zappa `handler.py` must be manually added to your Docker image, which I accomplish by grabbing the path with a simple Python command. This handler is then specified in the `CMD` setting, which causes it to run whenever a Docker container using this image is started. These steps are a must, since the `lambda_handler` function contains all the Zappa magic that routes API Gateway requests to the corresponding flask function or lets you execute raw Python commands in your function.
```

Comment:
```suggestion
The **only Zappa specific steps** are the last two. The Zappa `handler.py` must be manually added to your Docker image, which I accomplish by grabbing the path with a simple Python command. This handler is then specified in the `CMD` setting, which causes it to be run whenever a Docker container using this image is started. These steps are a must, since the `lambda_handler` function contains all the Zappa magic that routes API Gateway requests to the corresponding Flask function or lets you execute raw Python commands in your function.
```